group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -893,4 +1041,250 @@ public void testOnNewMetadataImage() {
+        assertEquals(mkSet("record1", "record2"), ctx.coordinator.records());
+        assertEquals(1, ctx.timer.size());
+
+        // Advance time to fire timer #2,
+        timer.advanceClock(10 + 1);
+
+        // Verify that the operation was executed.
+        assertEquals(mkSet("record1", "record2", "record3", "record4"), ctx.coordinator.records());
+        assertEquals(0, ctx.timer.size());
+    }
+
+    @Test
+    public void testRescheduleTimer() throws InterruptedException {
+        MockTimer timer = new MockTimer();
+        ManualEventProcessor processor = new ManualEventProcessor();
+        CoordinatorRuntime<MockCoordinator, String> runtime =
+            new CoordinatorRuntime.Builder<MockCoordinator, String>()
+                .withTime(timer.time)
+                .withTimer(timer)
+                .withLoader(new MockCoordinatorLoader())
+                .withEventProcessor(processor)
+                .withPartitionWriter(new MockPartitionWriter())
+                .withCoordinatorBuilderSupplier(new MockCoordinatorBuilderSupplier())
+                .build();
+
+        // Loads the coordinator.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Poll twice to process the pending events related to the loading.
+        processor.poll();
+        processor.poll();
+
+        // Check initial state.
+        CoordinatorRuntime<MockCoordinator, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.timer.size());
+
+        // The processor should be empty.
+        assertEquals(0, processor.size());
+
+        // Timer #1.
+        ctx.coordinator.timer.schedule("timer-1", 10, TimeUnit.MILLISECONDS,
+            () -> Collections.singletonList("record1"));
+
+        // The coordinator timer should have one pending task.
+        assertEquals(1, ctx.timer.size());
+
+        // Advance time to fire the pending timer.
+        timer.advanceClock(10 + 1);
+
+        // An event should be waiting in the processor.
+        assertEquals(1, processor.size());
+
+        // Schedule a second timer with the same key.
+        ctx.coordinator.timer.schedule("timer-1", 10, TimeUnit.MILLISECONDS,
+            () -> Collections.singletonList("record2"));
+
+        // The coordinator timer should still have one pending task.
+        assertEquals(1, ctx.timer.size());
+
+        // Schedule a third timer with the same key.
+        ctx.coordinator.timer.schedule("timer-1", 10, TimeUnit.MILLISECONDS,
+            () -> Collections.singletonList("record3"));
+
+        // The coordinator timer should still have one pending task.
+        assertEquals(1, ctx.timer.size());
+
+        // Advance time to fire the pending timer.
+        timer.advanceClock(10 + 1);
+
+        // Another event should be waiting in the processor.
+        assertEquals(2, processor.size());
+
+        // Poll twice to execute the two pending events.
+        assertTrue(processor.poll());
```

Comment:
So the first one is the record1 operation, but we don't actually add it to the coordinator because it was cancelled by the other operations with the same key?

Reply:
That's correct.